File: leetcode/easy/228/228.go
package main

import "strconv"

func summaryRanges(nums []int) []string {
	if len(nums) == 0 {
		return []string{}
	}

	result := []string{}

	start := nums[0]
	end := nums[0]

	for i := 1; i < len(nums); i++ {
		if nums[i] == end+1 {
			end = nums[i]
		} else {
			result = append(result, formatRange(start, end))
			start = nums[i]
			end = nums[i]
		}
	}

	result = append(result, formatRange(start, end))

	return result
}

func formatRange(start int, end int) string {
	if start == end {
		return strconv.Itoa(start)
	}

	return strconv.Itoa(start) + "->" + strconv.Itoa(end)
}

File: leetcode/easy/228/228.rs
impl Solution {
    pub fn summary_ranges(nums: Vec<i32>) -> Vec<String> {
        if nums.is_empty() {
            return vec![];
        }

        let mut result: Vec<String> = Vec::new();

        let mut start = nums[0];
        let mut end = nums[0];

        for i in 1..nums.len() {
            if nums[i] == end + 1 {
                end = nums[i];
            } else {
                result.push(Self::format_range(start, end));
                start = nums[i];
                end = nums[i];
            }
        }

        result.push(Self::format_range(start, end));

        result
    }

    fn format_range(start: i32, end: i32) -> String {
        if start == end {
            start.to_string()
        } else {
            format!("{}->{}", start, end)
        }
    }
}
